### PR TITLE
Enable LLVM NVPTX back-end

### PR DESCRIPTION
**Pull request by maleadt**
(Assumed) non-controversial parts of #19302, enables the LLVM NVPTX back-end (1 MB, 3% file size increase for `libLLVM`) and adds some crucial patches related to JuliaGPU/CUDAnative.jl#13 (all of them merged already):
* [D9168](https://reviews.llvm.org/D9168): segfault in LLVM
* [D24300](https://reviews.llvm.org/D24300): adds lowering for some intrinsics 
* [D23597](https://reviews.llvm.org/D23597): dependency of D24300

### CODE REVIEW

File: deps/llvm.mk
Comment (tkelman) [inline deps/llvm.mk]: why is this separate from the above list?

File: deps/llvm.mk
Comment (tkelman) [inline deps/llvm.mk]: we don't usually indent this kind of thing in makefiles

**Comment by tkelman** [comment]
will need to be tested on the buildbots to make sure it builds everywhere, if it's going to be enabled on all platforms by default

File: deps/llvm.mk
Comment (maleadt) [inline deps/llvm.mk]: Put it there in my branch to avoid merge conflicts, will change.

File: deps/llvm.mk
Comment (maleadt) [inline deps/llvm.mk]: Fair enough.

**Comment by maleadt** [comment]
Travis is happy. Any other configurations to manually test this on?

**Comment by tkelman** [comment]
all of the buildbots

**Comment by maleadt** [comment]
Right, but those only run on `master` and I don't have an account to request builds on this PR's branch.
So how do I make them run?

**Comment by ViralBShah** [comment]
Perhaps @tkelman or @staticfloat can share the creds with you for kicking off builds.

**Comment by tkelman** [comment]
I triggered a set of builds of this branch, they're running now. Something unrelated is wrong on the mac buildbot, it's having python issues with building the docs.

**Comment by maleadt** [comment]
The windows bots are acting up:
```
llvm-3.7.1/lib/Target/NVPTX/TargetInfo/NVPTXTargetInfo.cpp:10:19: fatal error: NVPTX.h: No such file or directory
 #include "NVPTX.h"
                   ^
compilation terminated.
```

Not sure how that could happen, it seems like LLVM 3.7 doesn't build if the NVPTX target is enabled?
I'll have a closer look.

**Comment by ViralBShah** [comment]
I guess things would be easier if were already using llvm 3.9.

**Comment by ViralBShah** [comment]
It does seem that llvm 3.9 is awfully close - but are we going to move to it for 0.6? If not, since it is late in the cycle, should we merge this?

Having CUDANative.jl work with Julia 0.6 out of the box would be a nice feature to add, like we did ARM for 0.5.

**Comment by tkelman** [comment]
> LLVM 3.7 doesn't build if the NVPTX target is enabled

that's a no, we shouldn't merge this without upgrading llvm

**Comment by tkelman** [comment]
Actually we should probably get the patches here onto master before we upgrade, so they go into the homebrew and appveyor binaries (I'll need to rebuild the latter) and the CI build matches what you get in a source build.

**Comment by maleadt** [comment]
OK, I can split those off. I just rebased this branch on top of your LLVM 3.9 update PR to first see if NVPTX will build correctly on Windows.

**Comment by tkelman** [comment]
Sure, worth testing. Do you have a login to the buildbots, or need me to fire that off?

**Comment by maleadt** [comment]
Shouldn't we also include the NVPTX back-end in the homebrew/appveyor binaries, if we want to ship that as part of Julia 0.6 @ LLVM 3.9? To do so, we could shield the `LLVM_TARGETS += NVPTX` behind a check against 3.9+ and merge this PR (if it builds correctly on all platforms).

That said, unless I still manage to create a PR for packaging an additional 'dev tools' tarball (with `llvm-config`, `include/llvm/*`, etc), users will have to build from source anyway.

**Comment by tkelman** [comment]
> users will have to build from source anyway

Sounds like it wouldn't add all that much value to turn on by default as long as this is the case?

**Comment by maleadt** [comment]
Ah yes, this will fail until staticfloat/julia-buildbot#54 is resolved of course.

```
-- Configuring incomplete, errors occurred!
CMake Error at CMakeLists.txt:3 (cmake_minimum_required):
  CMake 3.4.3 or higher is required.  You are running version 3.3.2
```

**Comment by ViralBShah** [comment]
Yes, if it builds on all platforms, this is a great idea.

**Comment by tkelman** [comment]
> we could shield the `LLVM_TARGETS += NVPTX` behind a check against 3.9+ and merge this PR (if it builds correctly on all platforms).

Let's do that then - just your commit, modified to only build the nvptx backend when using llvm >= 3.9. This branch seems to be working everywhere, but let's rearrange it so we can merge it independently. I'll need to rebase #19678 on top of #19681 anyway, and it'll have to wait for the homebrew formula and some benchmarking.

**Comment by tkelman** [comment]
> just your commit

Take mine out - can't merge this while 7851f775434af7df8d1acbb18927f13bc6229ec5 is in here

File: deps/llvm.mk
Comment (tkelman) [inline deps/llvm.mk]: do we need to worry about this case? it's not possible to use configure with 3.9 or newer, so I'd simplify this as something like
```
ifeq ($(LLVM_VER_SHORT),$(filter $(LLVM_VER_SHORT),3.3 3.4 3.5 3.6 3.7 3.8))
LLVM_TARGET_STRING := host
else
LLVM_TARGET_STRING := host;NVPTX
endif
```
if that works

Comment (maleadt) [inline deps/llvm.mk]: I figured I'd keep it a little more extensible, but sure that works too. I'll update the PR.

Comment (tkelman) [inline deps/llvm.mk]: people can always override `LLVM_TARGETS` (or `LLVM_TARGET_STRING` or whatever) in their `Make.user`, right?

Comment (maleadt) [inline deps/llvm.mk]: pretty much, yeah.

Comment (tkelman) [inline deps/llvm.mk]: then I guess we could keep it called `LLVM_TARGETS` just for minimal disruption to anyone who might already be doing this?

**Review by tkelman — APPROVED** [review]

File: deps/llvm.mk
Comment (singam-sanjay) [inline deps/llvm.mk]: Does this condition evaluate to true when LLVM_VER_SHORT is svn ?

Comment (yuyichao) [inline deps/llvm.mk]: It shouldn't.

Comment (singam-sanjay) [inline deps/llvm.mk]: Then how can I enable the NVPTX back-end for svn builds ?

Comment (yuyichao) [inline deps/llvm.mk]: It should be enabled since the false branch enables nvptx backend............

**Pull request merged**
